File: evaluation/S3/javascript/Medium/638-group-anagrams.js
// Source: https://leetcode.com/problems/group-anagrams/   |   Difficulty: Medium
//
// Problem Description:
// Given an array of strings strs, group the anagrams together. You can return the answer in any order.
//
// Example:
// Input: strs = ["eat","tea","tan","ate","nat","bat"]
//
// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
//
// Explanation:
//
//
// 	There is no string in strs that can be rearranged to form "bat".
// 	The strings "nat" and "tan" are anagrams as they can be rearranged to form each other.
// 	The strings "ate", "eat", and "tea" are anagrams as they can be rearranged to form each other.
//
// Constraints:
// 1 <= strs.length <= 104
// 	0 <= strs[i].length <= 100
// 	strs[i] consists of lowercase English letters.
//
// Helpful references (internal KB):
// - String Sorts (array, string, number, radix-sort, counting-sort)
//   • When to use: Use when sorting a large collection of fixed-length strings or integers, especially when the key range (alphabet size) is small. It is efficient for data with uniform key lengths.
//   • Idea: LSD Radix Sort sorts elements by processing digits or characters from the least significant to the most significant, using a stable counting sort for each position. It achieves a time complexity of O(W * (N + R)), where W is the key length, N is the number of items, and R is the radix.
//   • Invariants: After each pass 'd' (from 'w-1' down to '0'), the array 'a' is stably sorted based on characters from position 'd' to 'w-1'.; The 'count' array accurately reflects the frequency of each character at the current digit 'd' before cumulation.
//   • Tips: Utilize counting sort as the stable inner sorting algorithm for each digit pass.; Process characters or digits from right to left (least significant to most significant).
//   • Pitfalls: Inefficient for variable-length strings without proper padding or specialized handling.; Can be slower than comparison sorts for small input sizes or very large radix (R).
// - String Sorts (array, string, number, radix-sort, recursion, counting-sort)
//   • When to use: Use when sorting a large collection of strings or integers with a limited range of character/digit values. It is particularly efficient for fixed-length keys but can be adapted for variable lengths.
//   • Idea: Radix sort is a non-comparative integer sorting algorithm that sorts data with integer keys by grouping keys by individual digits or characters which share the same significant position and value. It typically achieves O(N*W) time complexity, where N is the number of items and W is the key length.
//   • Invariants: After each pass `d`, the elements are stably sorted according to their digits up to position `d`.; The counting sort subroutine correctly places elements into their sorted positions for the current digit, preserving relative order.
//   • Tips: LSD (Least Significant Digit) is simpler for fixed-length keys.; MSD (Most Significant Digit) is better for variable-length keys or when keys are very long.
//   • Pitfalls: Not efficient for small N or very large W due to constant factors.; Requires extra space for auxiliary array and counting array.
// - Suffix Array (string, array, counting-sort, radix-sort)
//   • When to use: Use when you need to efficiently query or analyze all suffixes of a string, often for pattern matching, string searching, or finding repeated substrings. It provides a sorted order of all suffixes.
//   • Idea: The Suffix Array algorithm constructs an array containing the starting indices of all suffixes of a string, sorted lexicographically. It typically uses a doubling strategy with counting sort or radix sort in each step, achieving an O(N log N) time complexity.
//   • Invariants: After iteration k, p contains the sorted indices of all cyclic substrings of length 2^k.; After iteration k, c[i] represents the equivalence class of the cyclic substring starting at i with length 2^k.
//   • Tips: Append a unique, smallest character (e.g., '$') to the string to handle suffixes as cyclic shifts.; Utilize counting sort for efficient sorting of characters and pairs of equivalence classes.
//   • Pitfalls: Incorrectly handling the sentinel character can lead to incorrect suffix order or out-of-bounds access.; Off-by-one errors when calculating indices for cyclic shifts or substring lengths.
//
// Notes:
// - Keep the given function signature unchanged.
// - Do not add custom input parsing or console I/O unless the template explicitly requires it; rely on parameters/return value only.
// - Use standard ECMAScript features supported by LeetCode (Node.js runtime); no external packages.
// - Be careful with number precision: JavaScript Number is double; use BigInt only when necessary and keep return type consistent with the problem requirements.
// - If you introduce helper functions, define them in the same file and avoid relying on global mutable state across test cases.

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const map = new Map();

    for (const str of strs) {
        // Sort the characters of the string to form the key
        const key = str.split('').sort().join('');
        
        if (!map.has(key)) {
            map.set(key, []); // Initialize the group if it doesn't exist
        }
        map.get(key).push(str); // Add the original string to the corresponding group
    }

    return Array.from(map.values()); // Return the grouped anagrams as an array of arrays
};

File: evaluation/S0/c/Medium/638-group-anagrams.c
// Source: https://leetcode.com/problems/group-anagrams/   |   Difficulty: Medium
//
// Problem Description:
// Given an array of strings strs, group the anagrams together. You can return the answer in any order.
//
// Example:
// Input: strs = ["eat","tea","tan","ate","nat","bat"]
//
// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
//
// Explanation:
//
//
// 	There is no string in strs that can be rearranged to form "bat".
// 	The strings "nat" and "tan" are anagrams as they can be rearranged to form each other.
// 	The strings "ate", "eat", and "tea" are anagrams as they can be rearranged to form each other.
//
// Constraints:
// 1 <= strs.length <= 104
// 	0 <= strs[i].length <= 100
// 	strs[i] consists of lowercase English letters.
//

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
char*** groupAnagrams(char** strs, int strsSize, int* returnSize, int** returnColumnSizes) {
    #define HASH_SIZE 10007

    typedef struct Node {
        char *str;
        struct Node *next;
    } Node;

    Node *hashTable[HASH_SIZE] = {0};
    char ***result = malloc(strsSize * sizeof(char **));
    *returnColumnSizes = malloc(strsSize * sizeof(int));
    *returnSize = 0;

    for (int i = 0; i < strsSize; i++) {
        char *str = strs[i];
        int count[26] = {0};
        for (char *p = str; *p; p++) {
            count[*p - 'a']++;
        }

        unsigned long hash = 5381;
        for (int j = 0; j < 26; j++) {
            hash = ((hash << 5) + hash) + count[j];
        }
        hash %= HASH_SIZE;

        Node *node = hashTable[hash];
        while (node) {
            if (strcmp(node->str, str) == 0) break;
            node = node->next;
        }

        if (!node) {
            node = malloc(sizeof(Node));
            node->str = str;
            node->next = hashTable[hash];
            hashTable[hash] = node;

            result[*returnSize] = malloc(strsSize * sizeof(char *));
            result[*returnSize][0] = str;
            (*returnColumnSizes)[*returnSize] = 1;
            (*returnSize)++;
        } else {
            for (int j = 0; j < *returnSize; j++) {
                if (strcmp(result[j][0], node->str) == 0) {
                    int colSize = (*returnColumnSizes)[j];
                    result[j][colSize] = str;
                    (*returnColumnSizes)[j]++;
                    break;
                }
            }
        }
    }

    return result;      
}
